eessi/testsuite/hooks.py

```diff
@@ -282,16 +282,50 @@ def _assign_one_task_per_gpu(test: rfm.RegressionTest):
     log(f'num_cpus_per_task set to {test.num_cpus_per_task}')
     log(f'num_tasks set to {test.num_tasks}')
 
+def _set_or_append_valid_systems(test: rfm.RegressionTest, valid_systems: str):
+    """
+    Sets test.valid_systems based on the valid_systems argument.
+    - When valid_systems is an empty string, test.valid_systems will be set equal to eessi.testsuite.constants.INVALID_SYSTEM
+    - When no test.valid_system was set yet, or it was at the default value ['*'], it will be overwritten by [valid_system]
+    - When a test.valid_system was already set, valid_system will be appended to it. This allows adding requests for multiple partition features by different hooks.
+    """
+
+    # This indicates an invalid test that always has to be filtered
+    if valid_systems == '':
+        test.valid_systems = [INVALID_SYSTEM]
+        return
+
+    # test.valid_systems wasn't set yet, so set it
+    if len(test.valid_systems) == 0:
+        test.valid_systems = [valid_systems]
```

Comment:
`valid_systems` is a required field, which means that it is always set, and we are now setting it to `['*']` in the class attribute. this means we can never have 0 items, and we can remove the first `if` condition.

Reply:
I wasn't sure if it couldn't be set to `[]` on the command line, but it seems not (at least I wasn't able to do it). Nevertheless: while we now have `[*]`as default valid systems by convention, this is something we set in the test class itself. If someone creates a test class with `[]` as default (which can easily happen), at least the current hook still works. 

I guess it depends on how 'hard' we want to enforce `['*']` as default. If we really want to enforce that, having this hook fail _could_ be one way to figure it out. But then the `== 0` case would just have to lead to a hard error or something.

What do you think, do we want to be that strict? Or do you prefer to keep the current version, which would _also_ make the `[]` default work?

Reply:
i’m thinking of the case where someone sets it to `[]` somewhere outside the class attributes, meaning that the test should not run due to some condition. in that case we should respect that, and not override it or fail. so i think it should always be `['*']` in the class attributes, but we should not enforce it in this hook.

i don’t think it is a big problem if someone sets it accidentally to `[]` in the class attribute, because in that case no test will run, so it’s clear something is wrong with the test.

if we really want to enforce it, a solution could be to set it to `['*']` in a class that inherits from `RegressionMixin` and make sure every test class inherits from both `RunOnlyRegressionTest` and the custom `RegressionMixin` child class, but of course then the question is how to enforce that, and maybe it makes things overly complex.
see here for an example: https://reframe-hpc.readthedocs.io/en/stable/tutorial_advanced.html#grouping-parameter-packs

Reply:
> i don’t think it is a big problem if someone sets it accidentally to [] in the class attribute, because in that case no test will run, so it’s clear something is wrong with the test.

That's a good point actually

> if we really want to enforce it, a solution could be to set it to ['*'] in a class that inherits from RegressionMixin and make sure every test class inherits from both RunOnlyRegressionTest and the custom RegressionMixin child class, but of course then the question is how to enforce that, and maybe it makes things overly complex.
see here for an example: https://reframe-hpc.readthedocs.io/en/stable/tutorial_advanced.html#grouping-parameter-packs

Yeah, I also thought maybe we should use Mixin classes more, after hearing e.g. Vasileios mention he also did that for his test library. I guess it might avoid more code duplication compared to the hooks and allows you to do default stuff like this. Enforcing it could be reasonably simple in the CI of the test suite by parsing the test file (it has to contain something like `class classname(..., my_mixin_class)`. Well, not something I'm going to dive into now, but it might make things cleaner. If we want to move in that direction, we should do it _before_ we have a million tests...